dev-tools/reconfigurator-cli/src/test_utils.rs

```diff
@@ -163,15 +210,139 @@ impl ReconfiguratorCliTestState {
+        description: &str,
+        f: F,
+    ) -> anyhow::Result<T>
+    where
+        F: FnOnce(&mut CollectionBuilder) -> anyhow::Result<T>,
+    {
+        self.change_state(description, |state| {
+            let mut builder = state.to_collection_builder()?;
+            let result = f(&mut builder)?;
+            state.system_mut().add_collection(builder.build())?;
+            Ok(result)
+        })
+    }
+
+    /// State change helper: create a new latest inventory collection, then pass
+    /// it to `f` which is allowed to edit it in abnormal ways that cannot be
+    /// done via `CollectionBuilder`.
+    pub fn inventory_edit_latest_low_level<F, T>(
+        &mut self,
+        description: &str,
+        f: F,
+    ) -> anyhow::Result<T>
+    where
+        F: FnOnce(&mut Collection) -> anyhow::Result<T>,
+    {
+        self.change_state(description, |state| {
+            let mut collection = state.to_collection_builder()?.build();
+            let result = f(&mut collection)?;
+            state.system_mut().add_collection(collection)?;
+            Ok(result)
+        })
+    }
+
+    /// State change helper: set the Clickhouse cluster policy.
+    pub fn set_clickhouse_policy(
+        &mut self,
+        description: &str,
+        policy: ClickhousePolicy,
+    ) {
+        self.change_description(description, |desc| {
+            desc.clickhouse_policy(policy);
+            Ok(())
+        })
+        .expect("closure can't fail");
+    }
+
+    /// State change helper: edit the latest blueprint, inserting a new latest
+    /// blueprint.
+    pub fn blueprint_edit_latest<F>(
+        &mut self,
+        description: &str,
+        f: F,
+    ) -> anyhow::Result<Arc<Blueprint>>
+    where
+        F: FnOnce(&mut BlueprintBuilder<'_>) -> anyhow::Result<()>,
+    {
+        let log = self.sim.log.clone();
+        self.change_state(description, |state| {
+            let rng = state.rng_mut().next_planner_rng();
+            let system = state.system_mut();
+            let blueprint = system
+                .get_blueprint(
+                    &system.resolve_blueprint_id(BlueprintId::Latest),
+                )
+                .expect("always have a latest blueprint");
+            let mut builder = BlueprintBuilder::new_based_on(
+                &log,
+                blueprint,
+                "ReconfiguratorCliTestState",
+                rng,
+            )?;
+            f(&mut builder)?;
+            let blueprint = Arc::new(builder.build(BlueprintSource::Test));
+            system.add_blueprint(Arc::clone(&blueprint))?;
+            Ok(blueprint)
+        })
+    }
+
+    /// State change helper: create a new latest blueprint that is a clone of
+    /// the current latest blueprint (but with a new ID and correct parent ID),
+    /// then pass it to `f` which is allowed to edit it in abnormal ways that
+    /// cannot be done via `BlueprintBuilder`.
+    pub fn blueprint_edit_latest_low_level<F>(
```

Comment:
nitty suggestion: I'd call this and the inventory analog "raw" instead of "low_level".